util/flocks/src/frwlock/mod.rs

```diff
@@ -0,0 +1,200 @@
+pub struct FrwLock<T: AsFd> {
+    cell: UnsafeCell<T>,
+}
+
+impl<T: AsFd> FrwLock<T> {
+    pub fn new(file: T) -> Self {
+        Self {
+            cell: UnsafeCell::new(file),
+        }
+    }
+
+    pub(crate) fn try_write(&self) -> Result<FrwLockWriteGuard<T>, FrwLockError> {
+        let file = unsafe { &*self.cell.get() };
+        match flock(file, FlockOperation::NonBlockingLockExclusive) {
+            Ok(_) => {
+                Ok(FrwLockWriteGuard {
+                    data: self.cell.get(),
+                })
+            },
+            Err(rustix::io::Errno::WOULDBLOCK) => Err(FrwLockError::LockNotAvailable),
+            Err(e) => Err(FrwLockError::FileError(e.into())),
+        }
+    }
+
+    pub(crate) fn try_read(&self) -> Result<FrwLockReadGuard<T>, FrwLockError> {
+        let file = unsafe { &*self.cell.get() };
+        match flock(file, FlockOperation::NonBlockingLockShared) {
+            Ok(_) => {
+                Ok(FrwLockReadGuard {
+                    data: self.cell.get(),
+                })
+            },
+            Err(rustix::io::Errno::WOULDBLOCK) => Err(FrwLockError::LockNotAvailable),
+            Err(e) => Err(FrwLockError::FileError(e.into())),
+        }
+    }
+
+    pub async fn write(&self) -> Result<FrwLockWriteGuard<T>, FrwLockError> {
+        loop {
+            match self.try_write() {
+                Ok(guard) => return Ok(guard),
+                Err(FrwLockError::LockNotAvailable) => {
+                    // Yield control to other tasks
+                    yield_now().await;
+                },
+                Err(e) => return Err(e),
+            }
+        }
+    }
+
+    pub async fn read(&self) -> Result<FrwLockReadGuard<T>, FrwLockError> {
+        loop {
+            match self.try_read() {
+                Ok(guard) => return Ok(guard),
+                Err(FrwLockError::LockNotAvailable) => {
+                    // Yield control to other tasks
+                    yield_now().await;
+                },
+                Err(e) => return Err(e),
+            }
+        }
+    }
+
+}
+
+// As long as T: Send + Sync, it's fine to send and share FrwLock<T> between threads.
+// If T were not Send, sending and sharing a FrwLock<T> would be bad, since you can access T through
+// FrwLock<T>.
+unsafe impl<T> Send for FrwLock<T> where T: AsFd + Sized + Send {}
+unsafe impl<T> Sync for FrwLock<T> where T: AsFd + Sized + Send + Sync {}
+// NB: These impls need to be explicit since we're storing a raw pointer.
+// Safety: Stores a raw pointer to `T`, so if `T` is `Sync`, the lock guard over
+// `T` is `Send`.
+unsafe impl<T> Send for FrwLockReadGuard<T> where T: AsFd + Sized + Sync {}
+unsafe impl<T> Sync for FrwLockReadGuard<T> where T: AsFd + Sized + Send + Sync {}
+unsafe impl<T> Send for FrwLockWriteGuard<T> where T: AsFd + Sized + Sync {}
+unsafe impl<T> Sync for FrwLockWriteGuard<T> where T: AsFd + Sized + Send + Sync {}
```

Comment:
I understand its fine to have these, as long as these invariants hold up. But I still think this is too heavy for the problem we are trying to solve, that is, some kind of dynamic state where different services can read or write vals. Why does it have to be a file and not some other "place" in memory?

Reply:
We shouldn't need to be handling thread safety and unsafe impls for simple config initialization. Just seems to much to me.

Reply:
Process A starts writing to the config.

Process B attempts to read whilst Process A is writing. 

Process B reads an invalid config. Process B crashes. 

Process A finishes writing.